constraint/pkg/client/drivers/local/local.go

```diff
@@ -75,22 +86,104 @@ func New(args ...Arg) drivers.Driver {
 	for _, arg := range args {
 		arg(d)
 	}
+
+	// adding externaldata builtin otherwise capabilities get overridden
+	// if a capability, like http.send, is disabled
+	if d.providerCache != nil {
+		d.capabilities.Builtins = append(d.capabilities.Builtins, &ast.Builtin{
+			Name: "external_data",
+			Decl: opatypes.NewFunction(opatypes.Args(opatypes.A), opatypes.A),
+		})
+	}
 	d.compiler.WithCapabilities(d.capabilities)
 	return d
 }
 
 var _ drivers.Driver = &driver{}
 
 type driver struct {
-	modulesMux   sync.RWMutex
-	compiler     *ast.Compiler
-	modules      map[string]*ast.Module
-	storage      storage.Store
-	capabilities *ast.Capabilities
-	traceEnabled bool
+	modulesMux    sync.RWMutex
+	compiler      *ast.Compiler
+	modules       map[string]*ast.Module
+	storage       storage.Store
+	capabilities  *ast.Capabilities
+	traceEnabled  bool
+	providerCache *externaldata.ProviderCache
 }
 
 func (d *driver) Init(ctx context.Context) error {
+	if d.providerCache != nil {
+		rego.RegisterBuiltin1(
+			&rego.Function{
+				Name:    "external_data",
+				Decl:    opatypes.NewFunction(opatypes.Args(opatypes.A), opatypes.A),
+				Memoize: true,
+			},
+			func(bctx rego.BuiltinContext, regorequest *ast.Term) (*ast.Term, error) {
+				var regoReq externaldata.RegoRequest
+				if err := ast.As(regorequest.Value, &regoReq); err != nil {
+					return nil, err
+				}
+				// only primitive types are allowed for keys
+				for _, key := range regoReq.Keys {
+					switch v := key.(type) {
+					case int:
+					case string:
+					case float64:
+					case float32:
+						break
+					default:
+						return nil, fmt.Errorf("type %v is not supported in external_data", v)
+					}
+				}
+
+				provider, err := d.providerCache.Get(regoReq.ProviderName)
+				if err != nil {
+					return nil, fmt.Errorf("unable to retrieve provider %v from cache", regoReq.ProviderName)
```

Comment:
What happens when a builtin returns an error?

Reply:
Does Rego panic or return the error to the user? Which behavior to we want?

Reply:
So I am talking to Torin about this. Looks like OPA changed behavior in 0.25:
https://github.com/open-policy-agent/opa/blob/main/CHANGELOG.md#0250

We can either set `rego.StrictBuiltinErrors(true)` or custom function can return `topdown.Halt`. However, looks like topdown.Halt is broken until v0.31 https://github.com/open-policy-agent/opa/issues/3534

Reply:
Can we sidestep the whole issue for most errors?

If we consider "missing provider" and "error calling provider" as system errors and report them that way in `RegoResponse` accordingly, there will be no need to invoke OPA error behavior.

Other errors, like parsing errors, seem a bit more borderline, but I think we can still consider them "system errors"

Reply:
Do we treat all errors the same in terms of fail open vs fail close?

Reply:
similarly, what happens if there is an error when sending to request back to provider?

Reply:
updated. returning `RegoResponse` for most errors